### PR TITLE
POC: Add option to build/push once `prefect deploy`

### PR DESCRIPTION
**Pull request by serinamarie**
Users often have multiple deployments to deploy at once, e.g. `prefect deploy -n deployment1 -n deployment2` or `prefect deploy --all`. However, they often have a single `build` action defined at the project level, and generally a single Dockerfile. Currently the build and push actions are performed for every deployment, leading to x number of images built and pushed for x number of deployments, even when they are identical.

It would be ideal to have the option to only run those steps once, with one Docker image built and pushed. 

### Checklist

- [x] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] This pull request includes tests or only affects documentation.
- [x] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`, `docs`.
  

For documentation changes:

- [ ] This pull request includes redirect settings in `netlify.toml` for files that are removed or renamed

### CODE REVIEW

File: src/prefect/cli/deploy.py [hunk 1]
```diff
@@ -250,6 +258,7 @@ async def deploy(
                 actions=actions,
                 deploy_all=deploy_all,
                 ci=ci,
+                build_push_once=build_push_once,
```
Comment (serinamarie) [inline src/prefect/cli/deploy.py:261]: The option is passed into `prefect deploy` here, which passes it to `_run_multi_deploy` at line 686

File: src/prefect/cli/deploy.py [hunk 2]
```diff
@@ -699,7 +716,17 @@ async def _run_multi_deploy(
...
+        await _run_single_deploy(
+            deploy_config,
+            actions,
+            ci=ci,
+            build_push_once=build_push_once,
+            image_has_been_built=image_has_been_built,
+        )
+
+        if build_push_once and not image_has_been_built:
+            image_has_been_built = True
```
Comment (serinamarie) [inline src/prefect/cli/deploy.py:729]: Right now I handle tracking if we should build the image once and if we have built it, outside of `_run_single_deploy`, as currently it looks is something like:

```python
def deploy():
    _run_multi_deploy()

def _run_multi_deploy():
    for d in deployment:
        _run_single_deploy()
```

so with the change if would look like this:
```python
def deploy():
    _run_multi_deploy()

def _run_multi_deploy():
    image_has_been_built = False
    for d in deployment:
        _run_single_deploy(build_push_once, image_has_been_built)
        if build_push_once and not image_has_been_built:
            image_has_been_built = True
```

File: src/prefect/cli/deploy.py [hunk 3]
```diff
@@ -194,6 +194,14 @@ async def deploy(
             " provided."
         ),
     ),
+    build_push_once: bool = typer.Option(
+        False,
+        "--build-push-once",
+        help=(
```
Comment (serinamarie) [inline src/prefect/cli/deploy.py:200]: I would appreciate feedback on whether it should be a single flag like `--build-push-once`, `--build once` or something else :)

File: src/prefect/cli/deploy.py [hunk 4]
```diff
@@ -694,7 +711,17 @@
                 app.console.print("Skipping unnamed deployment.", style="yellow")
                 continue
         app.console.print(Panel(f"Deploying {deploy_config['name']}", style="blue"))
-        await _run_single_deploy(deploy_config, actions, ci=ci)
+
```
[the remark this answers is not in the document]
Comment (serinamarie) [reply src/prefect/cli/deploy.py:718]: not new to this PR

**Comment by desertaxle** [comment]
How would the `--build-once` flag work with deployments that have a specific build action that overrides a general build action? It looks like it might skip specific build steps in that case. I think we should explore some additional approaches to see if there's a more tactical way to handle action optimization. Makefile's method of caching could be a good approach to emulate.

**Comment by serinamarie** [comment]
> How would the --build-once flag work with deployments that have a specific build action that overrides a general build action? It looks like it might skip specific build steps in that case. I think we should explore some additional approaches to see if there's a more tactical way to handle action optimization. Makefile's method of caching could be a good approach to emulate.

@desertaxle The way it's written now, it would not. It's only geared toward those who want to use the same build/push step project wide. It wouldn't be hard to handle overrides though. However, I'll look into the approach you described.

**Comment by desertaxle** [comment]
Both of those could definitely work. I'd expect different users will want different behavior, so having an eye towards configurability might be prudent. I think it might be a good idea to scope this optimization to the Docker build and push steps for now. That would allow us to experiment with step run optimization while keeping the option for general step run optimization in the future. We could read existing images to determine if an step should run as a substitute for writing information to a new file.

Let's find some time next week to spike out some options. I bet we can land on a good approach after a bit of experimentation.

**Comment by serinamarie** [comment]
@desertaxle, sounds great, appreciate it!

**Comment by serinamarie** [comment]
After discussion, @desertaxle and I are deferring this until we have more information on usage!

**Pull request closed without merging**